hail/src/main/scala/is/hail/expr/types/physical/PCanonicalStruct.scala

```diff
@@ -0,0 +1,253 @@
+package is.hail.expr.types.physical
+
+import is.hail.annotations.{CodeOrdering, Region, UnsafeInserter}
+import is.hail.asm4s.Code
+import is.hail.expr.ir.{EmitMethodBuilder, SortOrder}
+import is.hail.utils.{ArrayBuilder, fatal, plural, prettyIdentifier}
+import org.apache.spark.sql.Row
+import is.hail.utils._
+
+object PCanonicalStruct {
+  private val requiredEmpty = PCanonicalStruct(Array.empty[PField], true)
+  private val optionalEmpty = PCanonicalStruct(Array.empty[PField], false)
+
+  def empty(required: Boolean = false): PStruct = if (required) requiredEmpty else optionalEmpty
+
+  def apply(required: Boolean, args: (String, PType)*): PStruct =
+    PCanonicalStruct(args
+      .iterator
+      .zipWithIndex
+      .map { case ((n, t), i) => PField(n, t, i) }
+      .toFastIndexedSeq,
+      required)
+
+  def apply(args: (String, PType)*): PStruct =
+    PCanonicalStruct(false, args:_*)
+}
+
+final case class PCanonicalStruct(fields: IndexedSeq[PField], required: Boolean = false) extends PStruct with PCanonicalBaseStruct {
+  assert(fields.zipWithIndex.forall { case (f, i) => f.index == i })
+
+  val types: Array[PType] = fields.map(_.typ).toArray
+
+  if (!fieldNames.areDistinct()) {
+    val duplicates = fieldNames.duplicates()
+    fatal(s"cannot create struct with duplicate ${plural(duplicates.size, "field")}: " +
+      s"${fieldNames.map(prettyIdentifier).mkString(", ")}", fieldNames.duplicates())
+  }
+
+  def copy(fields: IndexedSeq[PField] = this.fields, required: Boolean = this.required): PStruct = PCanonicalStruct(fields, required)
+
+  override def truncate(newSize: Int): PStruct =
+    PCanonicalStruct(fields.take(newSize), required)
+
+  override def codeOrdering(mb: EmitMethodBuilder, other: PType): CodeOrdering =
```

Comment:
aren't these already defined? Might want to make them `final` to prevent this error.

Reply:
yeah I fixed this around the same time you commented. will add final